### PR TITLE
TDL-19101: Add custom exception handling

### PR DESCRIPTION
**Pull request by savan-chovatiya**
# Description of change
[TDL-19101](https://jira.talendforge.org/browse/TDL-19101): Add custom exception handling and error logging
- Added customer exception codes map to handle the exception and custom message if message not found in API response.
- Kept retrying mechanism as it is for 4xx and 5xx.

# Manual QA steps
 - Verified that exception is raised with a proper error message if a message is present in API response for error.
- Verified that exception is raised with a custom message if API response does not have a message.

 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_chargebee/client.py [hunk 1]
```diff
@@ -12,13 +12,122 @@
...
+
+
+STATUS_CODE_EXCEPTION_MAPPING = {
+    400: {
+        "raise_exception": ChargebeeBadRequestError,
+        "message": "The request URI does not match the APIs in the system.",
+    },
+    401: {
+        "raise_exception": ChargebeeUnauthorizedError,
+        "message": "The user is not authorized to use the API.",
```
Comment (NevilParikh14) [inline tap_chargebee/client.py:62]:
```suggestion
        "raise_exception": ChargebeeAuthenticationError,
        "message": "The user is not authenticated to use the API.",
```
Ref: https://apidocs.chargebee.com/docs/api?prod_cat_ver=2#error_codes_list

File: tap_chargebee/client.py [hunk 2]
```diff
@@ -12,13 +12,122 @@
...
+    exception = STATUS_CODE_EXCEPTION_MAPPING.get(status_code, {}).get(
+                "raise_exception")
+    # If exception is not mapped for any code then use Server4xxError and Server5xxError respectively
+    if not exception:
+        if status_code > 400 and status_code < 500:
+            exception = Server4xxError
+        elif status_code > 500:
+            exception = Server5xxError
+        else:
+            exception = ChargebeeError
```
Comment (NevilParikh14) [inline tap_chargebee/client.py:108]: Can you please add a unit test case for these else condition?

Comment (savan-chovatiya) [reply tap_chargebee/client.py:108]: Added unit tets to verify above else condition too.

File: tap_chargebee/client.py [hunk 3]
```diff
@@ -70,19 +179,9 @@ def make_request(self, url, method, params=None, body=None):
...
-            }
-
-        if response.status_code == 429:
-            raise Server429Error()
+        if response.status_code != 200:
+            raise_for_error(response)
 
-        if response.status_code >= 400:
-            raise Server4xxError(response_json)
+        response_json = response.json()
```
Comment (harshpatel4crest) [inline tap_chargebee/client.py:185]: We directly return `response.json()`

Comment (savan-chovatiya) [reply tap_chargebee/client.py]: Yes, updated code to return response.json() directly.

**Review by harshpatel4crest — APPROVED** [review]

File: tap_chargebee/client.py [hunk 4]
```diff
@@ -12,14 +12,123 @@
...
+class ChargebeeNotFoundError(Server4xxError):
+    pass
+
+class ChargebeeMethodNotAllowedError(Server4xxError):
+    pass
+
+class ChargebeeNotProcessedError(Server4xxError):
+    pass
+
+class ChargebeeRateLimitError(Server429Error):
```
Comment (dbshah1212) [inline tap_chargebee/client.py:45]: Passing 429 seems redundant, we can pass 4xx here

Comment (savan-chovatiya) [reply tap_chargebee/client.py]: Removed Server429Error and pass Server4xxError as suggested.

**Review by dbshah1212 — APPROVED** [review]

**Review by KrisPersonal — APPROVED** [review]

**Comment by RushT007** [comment]
Reviewed, will wait for CCi to run successfully

**Comment by harshpatel4crest** [comment]
> Reviewed, will wait for CCi to run successfully

The build is passing now.

**Review by RushT007 — APPROVED** [review]

**Review by kspeer825 — APPROVED** [review]

**Pull request merged**